### PR TITLE
feat(win): new capture method, Windows.Graphics.Capture 

### LINKED ISSUE #832: Xbox Game Bar doesn't show up in the stream
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your issue described in the documentation?

- [X] I have read the documentation

### Is your issue present in the nightly release?

- [X] This issue is present in the nightly release

### Describe the Bug

The Xbox Game Bar isn't captured and sent to the client.

Steps to reproduce:
* Start streaming from a Windows host to any client
* Open the Xbox Game Bar (Win+G by default)

The Game Bar appears in the host, but not in the client.

### Expected Behavior

Same as GameStream: the Game Bar shows up in both, the host and the client.

### Additional Context

_No response_

### Host Operating System

Windows

### Operating System Version

11 (Build 22621.1105)

### Architecture

64 bit

### Sunshine commit or version

0.17.0

### Package

Windows - installer

### GPU Type

Nvidia

### GPU Model

GeForce GTX 1650

### GPU Driver/Mesa Version

528.24

### Capture Method (Linux Only)

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by tez011**
## Description
This adds a capture backend on Windows that uses the [Windows.Graphics.Capture API](https://learn.microsoft.com/en-us/windows/uwp/audio-video-camera/screen-capture). This is a more recent API that's capable of capturing the Xbox Game Bar, which the DXGI Desktop Duplication API seemingly cannot do.

This pull request depends on C++20 compiler support (for the Microsoft WinRT headers, which define this API) as well as Universal CRT runtime support. These have already been integrated into Sunshine in #2322 and #2323. 

See also #2149.

Issues Fixed or Closed

Fixes #832 (manually verified)

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency update (updates to dependencies)
- [x] Documentation update (changes to documentation)
- [x] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

## Branch Updates
LizardByte requires that branches be up-to-date before merging. This means that after any PR is merged, this branch
must be updated before it can be merged. You must also
[Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I want maintainers to keep my branch updated

### CODE REVIEW

**Comment by tez011** [comment]
Right now, this pull request does not contain changes to support Sunshine's service model.
Linking https://github.com/LizardByte/Sunshine/pull/2149#issuecomment-1955746301 for context.

File: .github/workflows/CI.yml [hunk 1]
```diff
@@ -942,6 +942,7 @@ jobs:
             git
             mingw-w64-ucrt-x86_64-boost
             mingw-w64-ucrt-x86_64-cmake
+            mingw-w64-ucrt-x86_64-cppwinrt
```
Comment (ReenigneArcher) [inline .github/workflows/CI.yml:945]: Please add this to the build docs and https://github.com/LizardByte/Sunshine/blob/7702e55476fef7edf7e421a3631dcde2686afe73/.codeql-prebuild-cpp-Windows.sh#L8

File: src_assets/common/assets/web/configs/tabs/Advanced.vue [hunk 2]
```diff
@@ -59,14 +59,22 @@ const config = ref(props.config)
     </div>
 
     <!-- Capture -->
-    <div class="mb-3" v-if="platform === 'linux'">
+    <div class="mb-3" v-if="platform === 'linux' || platform === 'windows'">
```
Comment (ReenigneArcher) [inline src_assets/common/assets/web/configs/tabs/Advanced.vue:62]:
```suggestion
    <div class="mb-3" v-if="platform !== 'macos'">
```

File: src_assets/common/assets/web/configs/tabs/Advanced.vue [hunk 3]
```diff
@@ -59,14 +59,22 @@ const config = ref(props.config)
...
+            <option value="ddx">Desktop Duplication API</option>
+            <option value="wgc">Windows.Graphics.Capture {{ $t('_common.beta') }}</option>
+          </template>
+          <template #linux>
+            <option value="nvfbc">NvFBC</option>
+            <option value="wlr">wlroots</option>
+            <option value="kms">KMS</option>
+            <option value="x11">X11</option>
+          </template>
+        </PlatformLayout>
```
Comment (ReenigneArcher) [inline src_assets/common/assets/web/configs/tabs/Advanced.vue:77]:
```suggestion
        <PlatformLayout :platform="platform">
          <template #linux>
            <option value="nvfbc">NvFBC</option>
            <option value="wlr">wlroots</option>
            <option value="kms">KMS</option>
            <option value="x11">X11</option>
          </template>
          <template #windows>
            <option value="ddx">Desktop Duplication API</option>
            <option value="wgc">Windows.Graphics.Capture {{ $t('_common.beta') }}</option>
          </template>
        </PlatformLayout>
```

Just alphabetizing the platforms.

**Review by ReenigneArcher — CHANGES_REQUESTED** [review]

File: .github/workflows/CI.yml (hunk 1 above)
Comment (tez011) [reply .github/workflows/CI.yml:945]: I must have forgotten to make changes to the codeql file in #2323; I added them here.

File: .codeql-prebuild-cpp-Windows.sh [hunk 4]
```diff
@@ -11,18 +11,20 @@ pacman --noconfirm -S \
...
-  mingw-w64-x86_64-curl \
-  mingw-w64-x86_64-miniupnpc \
-  mingw-w64-x86_64-nlohmann-json \
-  mingw-w64-x86_64-nodejs \
-  mingw-w64-x86_64-onevpl \
-  mingw-w64-x86_64-openssl \
-  mingw-w64-x86_64-opus \
-  mingw-w64-x86_64-rust \
-  mingw-w64-x86_64-toolchain
+  mingw-w64-ucrt-x86_64-boost \
```
Comment (ReenigneArcher) [inline .codeql-prebuild-cpp-Windows.sh:14]: Do you think ucrt will work if the setup step isn't using that? https://github.com/LizardByte/Sunshine/blob/7702e55476fef7edf7e421a3631dcde2686afe73/.github/workflows/codeql.yml#L142

If not, I might have to figure out how to handle this separately as we use the same workflow file across all repos. We could also change the other repo using msys2 to ucrt (and we should anyway).

Comment (ReenigneArcher) [reply .codeql-prebuild-cpp-Windows.sh:14]: Actually the other one is already using ucrt for the ci build, so I will convert codeql to use it as well.

https://github.com/LizardByte/libdisplaydevice/blob/ff361a83c4b2d97df25c33efba99125f2d04b3de/.github/workflows/ci.yml#L58

Comment (ReenigneArcher) [reply .codeql-prebuild-cpp-Windows.sh:14]: Can you apply these changes directly to this PR?

https://github.com/LizardByte/Sunshine/pull/2581/files

Comment (ReenigneArcher) [reply .codeql-prebuild-cpp-Windows.sh:14]: I made the updates here: https://github.com/LizardByte/Sunshine/pull/2581

Comment (tez011) [reply .codeql-prebuild-cpp-Windows.sh:14]: does `msystem: ucrt64` need to be added to https://github.com/LizardByte/Sunshine/blob/7702e55476fef7edf7e421a3631dcde2686afe73/.github/workflows/codeql.yml#L146 as well?

Comment (ReenigneArcher) [reply .codeql-prebuild-cpp-Windows.sh:14]: @tez011 I just merged that, so you should be good if you rebase.

**Comment by ReenigneArcher** [comment]
I've quickly tested this locally.

- CPU usage is < 4% while streaming with either capture method.
- RAM usage is < 5mb while streaming with either capture method.
- Can't use the new capture method when running it as a service, which I suppose is expected for now.
- Can start Sunshine from start menu shortcut.

https://github.com/LizardByte/Sunshine/pull/2320

Was there anything else broken when we tried to merge this before?

**Review by ReenigneArcher — APPROVED** [review]

**Comment by tez011** [comment]
I seem to remember a couple other things:

- Configuration UI comes up no slower than usual (was previously broken due to regex)
- HDR may still not work (I don't have hardware capable of testing that)

**Comment by ReenigneArcher** [comment]
@tez011

UI seems to be coming up quickly for me.

> HDR may still not work (I don't have hardware capable of testing that)

This was just with the new capture method, correct? The old one is unaffected?

**Comment by tez011** [comment]
> > HDR may still not work (I don't have hardware capable of testing that)
> 
> This was just with the new capture method, correct? The old one is unaffected?

Yes. I could never reproduce it, but I recall this discussion from the old pull request. https://github.com/LizardByte/Sunshine/pull/2149#issuecomment-2027437793

**Comment by ReenigneArcher** [comment]
Alright, this will be merged soon. Thank you.

**Comment by Kobain-Seo** [comment]
When i enabled this option, sunshine fails to start.

Logs:

[2024-06-20 14:24:29.019908] [0x00002e0c] [info]    Gamepad x360 is disabled due to gamepads.vigem-not-available
[2024-06-20 14:24:29.021147] [0x00002e0c] [info]    Gamepad ds4 is disabled due to gamepads.vigem-not-available
[2024-06-20 14:24:29.022127] [0x00002e0c] [info]    Gamepad x360 is disabled due to gamepads.vigem-not-available
[2024-06-20 14:24:29.023114] [0x00002e0c] [info]    Gamepad ds4 is disabled due to gamepads.vigem-not-available
[capture] -- [wgc]
[encoder] -- [nvenc]
[qp] -- [0]
[nvenc_twopass] -- [full_res]
[fec_percentage] -- [0]
[controller] -- [disabled]
[native_pen_touch] -- [disabled]
[nvenc_vbv_increase] -- [40]
[2024-06-20 14:24:29.024813] [0x00002e0c] [info]    Gamepad x360 is disabled due to gamepads.vigem-not-available
[2024-06-20 14:24:29.025803] [0x00002e0c] [info]    Gamepad ds4 is disabled due to gamepads.vigem-not-available
[2024:06:20:14:24:29]: Info: Sunshine version: v2024.620.13611
[2024:06:20:14:24:29]: Info: nvprefs: Opened undo file from previous improper termination
[2024:06:20:14:24:29]: Info: nvprefs: Restored OGL_CPL_PREFER_DXPRESENT for base profile
[2024:06:20:14:24:29]: Info: nvprefs: Restored global profile settings from undo file - deleting the file
[2024:06:20:14:24:29]: Info: nvprefs: No need to modify application profile settings
[2024:06:20:14:24:29]: Info: nvprefs: Changed OGL_CPL_PREFER_DXPRESENT to OGL_CPL_PREFER_DXPRESENT_PREFER_ENABLED for base profile
[2024:06:20:14:24:29]: Info: Compiling shaders...
[2024:06:20:14:24:29]: Info: System tray created
[2024:06:20:14:24:29]: Info: Compiled shaders
[2024:06:20:14:24:29]: Warning: Gamepad x360 is disabled due to gamepads.vigem-not-available
[2024:06:20:14:24:29]: Warning: Gamepad ds4 is disabled due to gamepads.vigem-not-available
[2024:06:20:14:24:29]: Warning: No gamepad input is available
[2024:06:20:14:24:29]: Info: Trying encoder [nvenc]
[2024:06:20:14:24:29]: Info: ddprobe.exe [1] [] returned: 0x00000000
[2024:06:20:14:24:29]: Info: Set GPU preference: 1
[2024:06:20:14:24:29]: Info:
Device Description : NVIDIA GeForce RTX 3060
Device Vendor ID   : 0x000010DE
Device Device ID   : 0x00002504
Device Video Mem   : 12115 MiB
Device Sys Mem     : 0 MiB
Share Sys Mem      : 8113 MiB
Feature Level      : 0x0000B100
Capture size       : 1920x1080
Offset             : 0x0
Virtual Desktop    : 1920x1080
[2024:06:20:14:24:29]: Info: Active GPU has HAGS enabled
[2024:06:20:14:24:29]: Info: Using realtime GPU priority
[2024:06:20:14:24:29]: Info:
Colorspace         : DXGI_COLOR_SPACE_RGB_FULL_G22_NONE_P709
Bits Per Color     : 8
Red Primary        : [0.639648,0.330078]
Green Primary      : [0.299805,0.599609]
Blue Primary       : [0.149414,0.0595703]
White Point        : [0.3125,0.329102]
Min Luminance      : 0.5 nits
Max Luminance      : 270 nits
Max Full Luminance : 270 nits

ddprobe failed, but i don't know why.

**Comment by ReenigneArcher** [comment]
You have to disable the service and start Sunshine manually.

**Comment by Kobain-Seo** [comment]
> You have to disable the service and start Sunshine manually.

I think it is not relevant to running manually. I tested it on Windows 10 22H2 and sunshine failed to run, and It worked on Windows 11 22H2. Is it normal it does not work on Windows 10 ?

**Comment by tez011** [comment]
Yes, that’s expected. This API is quite new.

**Comment by Kobain-Seo** [comment]
after patched #2825, It works on Windows 10, but have to be run manually, not by service.

**Comment by PieroPontra** [comment]
Latest version of Win11 but the new beta WinGraphicsCapture that I switched to manually in configs says it can't find any encoder or display attached. I already tried to manually enter names from dxci command but same results, any hints?

**Comment by WXZhao7** [comment]
I noticed from https://github.com/LizardByte/Sunshine/pull/2149#issuecomment-1955746301 that WGC should be able to capture the UAC secure desktop. I have tested several pre-release versions, including v2024.714.230159. Unfortunately, Moonlight lose control when the UAC window pops up. 
Is the capture of UAC prompts by WGC still incomplete, or is it not planned for implementation?

**Comment by roob0** [comment]
@PieroPontra You have to close the service and launch Sunshine from the exe, but so you can't stream the lockscreen. However, I created a simple solution for switching automatically from service to exe and from ddx wo wgc (and viceversa) here:
https://github.com/LizardByte/Sunshine/pull/2149

**Comment by karasuhebi** [comment]
Is there any other benefits to this capture method?

**Pull request merged**